addons/sale_stock_margin/tests/test_sale_stock_margin.py

```diff
@@ -410,3 +410,39 @@ def test_add_avco_product_on_delivery_cost_on_sale_order(self):
         delivery.move_ids.quantity = 10
         delivery.button_validate()
         self.assertEqual(sale_order.order_line.filtered(lambda sol: sol.product_id == product2).purchase_price, 10)
+
+    def test_purchase_price_recompute_after_cancel_pricelist_change(self):
+        """Test that purchase_price is recomputed when pricelist changes after order cancellation."""
+        currency_usd = self.env.ref('base.USD')
+        currency_eur = self.env.ref('base.EUR')
+        date = fields.Date.context_today(self.env.user)
+        self.env['res.currency.rate'].create([
+            {'currency_id': currency_usd.id, 'rate': 1.0, 'name': date, 'company_id': False},
+            {'currency_id': currency_eur.id, 'rate': 0.5, 'name': date, 'company_id': False},
+        ])
+        pricelist_usd = self.env['product.pricelist'].create({
+            'name': 'USD Pricelist',
+            'currency_id': currency_usd.id,
+        })
+        pricelist_eur = self.env['product.pricelist'].create({
+            'name': 'EUR Pricelist',
+            'currency_id': currency_eur.id,
+        })
```

Comment:
```suggestion
        pricelist_usd = self.env['product.pricelist'].create([
            {
                'name': 'USD Pricelist',
                'currency_id': currency_usd.id,
            },
            {
                'name': 'EUR Pricelist',
                'currency_id': currency_eur.id,
            },
        ])
```
Always batch `create()` whenever possible